### DISCUSSION TITLE
Talos for IoT & lite Edge - standalone Kubelet

### DISCUSSION THREAD

**Discussion by fullmetal-fred** [siderolabs/talos#8343] [Ideas]
Hey Sidero team!

I love Talos. I've been following this project for years now, and I think, for the record, you're on the path to becoming THE defacto way to run baremetal, on-prem, and self-managed in the cloud / multi-cloud k8s. Great work 🙌🏻

Yes, Talos as a declarative OS is PERFECT for IoT and Edge...but k8s is not.

K8s is just too heavy and too complicated for single-node IoT / Edge use-cases. The memory, cpu, and configuration overhead is unnecessary for single nodes which will never do more than run 2-3 of the same pods forever. It limits our ability to use more constrained compute. It's also too expensive to use Omni for this sort of thing...most solutions in this space cost $0.30-$1.50 per node per month. 

What we really need for that usecase is a single node container runtime, without a control plane, configured solely by the Talos config (using static pods via the Kubelet & containerd I imagine), also known as a standalone Kubelet. Machine management at scale would be done via a tool like Omni, or using a more use case-specific agent as an extension or a static pod, which would locally interact with apid. But the key here is that Pods could be deployed declaratively with config patches via the API. 

I know much of this (theoretically) works today within Talos. We can inform the Kubelet to run standalone. We can configure static pods via the machine config. But I assume this isn't a supported configuration and that Talos would be confused running in this way. Perhaps I am wrong?

So I ask: what would it take for a single node, standalone Kubelet configuration of Talos to be an "official" usecase? Are there configuration options that would need to be added to inform Talos that it's ok there's no control plane? Or is there a golden config that would support this today and merely needs to be documented for the use case?

I'd love to hear your thoughts, and thank you for your incredible work!! 🤘🏻

PS: for a comparison solution in this space, see Balena Engine / OS / Cloud.

PPS: Kelsey Hightower himself has a tutorial for running in a standalone Kubelet mode...you can find that here: https://github.com/kelseyhightower/standalone-kubelet-tutorial

**Comment by rothgar** [comment]
The main idea to get us to a standalone talos (no k8s) option would be to make the cluster section of the config optional and probably make a new `type` for machined to know how to handle maintenance. This would still read and configure the `machine` section of the machine config but wouldn't start all of the kubernetes specific parts (etcd, api server, etc). Instead you could configure the `mahcine.pods` section with static pods you want defined on the system.

This doesn't work today and currently isn't planned, but we know we want something like this for standalone etcd database nodes. The idea being you would still define a cluster section but only include the necessary etcd configuration (eg certificates).

These are both reasonable requests, but we don't have the time to do them and there has been no customer demand for them yet. Therefore, they have not been prioritized. If this is something the community would like to contribute we welcome design proposals and PRs.

**Reply by fullmetal-fred** [reply]
Thanks for the feedback @rothgar. Makes sense to me. I'm bummed more customers haven't requested this but I get it...it's those of us working in more purpose-built IoT/edge systems that are begging for things like this as an alternative to Yocto, Buildkit, Balena, and so on. 

I wish I understood the inner workings of Talos well enough to submit PRs to enable it. Maybe someday!

Thanks again for the response.
